### PR TITLE
Fix __shared__ complex<T> undefined behavior

### PR DESCRIPTION
**Pull request by nluehr**
std::complex<T> has a non-empty constructor (zero assignment) that is not
compatible with CUDA __shared__ memory. This fixes current reliance on
undefined behavior (and removes an unnecessary run-time initialization).

### CODE REVIEW

**Comment by ekelsen** [comment]
Does this have any performance implications?  Did you run the reduction benchmarks?

**Comment by nluehr** [comment]
Did not check performance of reductions, but changes are expected to slightly improve performance, if anything. With the old code the compiler was generating unnecessary instructions for each thread to independently zero the shared memory region.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ekelsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by ekelsen — APPROVED** [review]

**Comment by nluehr** [comment]
@tfboyd can you look into getting this merged?

File: tensorflow/core/kernels/reduction_gpu_kernels.cu.h [hunk 1]
```diff
@@ -244,6 +244,33 @@ __global__ void RowReduceKernel(
...
+  __host__ __device__ storage_type() {}
+  __host__ __device__ operator T1() { return val; }
+  __host__ __device__ storage_type<T1>& operator=(const T1& in) {
+    val = in;
+    return *this;
+  }
+};
+
+template <typename T2>
+struct storage_type<std::complex<T2>> {
```
Comment (zheng-xq) [inline tensorflow/core/kernels/reduction_gpu_kernels.cu.h]: A comment on why we have this specialization for std::complex

Comment (zheng-xq) [inline tensorflow/core/kernels/reduction_gpu_kernels.cu.h]: Is this change only relevant to shared memory? If so, maybe shared_mem_storage_type?

**Review by zheng-xq — COMMENTED** [review]
Also it will be great to double check that there is no performance regression. Thanks!

**Comment by zheng-xq** [comment]
@reedwm, please help move this PR through the merge process. Thanks!
